File: src/xai/sdk/typescript/src/clients/governance-client.ts
/**
 * Governance Client for XAI SDK
 *
 * Handles governance proposals, voting, and proposal lifecycle management.
 */

import { HTTPClient } from '../utils/http-client';
import {
  Proposal,
  ProposalStatus,
  ProposalQueryParams,
  CreateProposalParams,
  VoteChoice,
  PaginatedResponse,
} from '../types';
import { GovernanceError, ValidationError } from '../errors';

/**
 * Client for governance operations
 */
export class GovernanceClient {
  constructor(private httpClient: HTTPClient) {}

  /**
   * List governance proposals
   *
   * @param params - Query parameters with status filter, limit, and offset
   * @returns Paginated list of proposals
   *
   * @example
   * ```typescript
   * const result = await client.governance.listProposals({
   *   status: 'active',
   *   limit: 10
   * });
   * console.log('Active proposals:', result.data);
   * ```
   */
  async listProposals(params: ProposalQueryParams = {}): Promise<PaginatedResponse<Proposal>> {
    const limit = Math.min(params.limit || 20, 100);
    const offset = params.offset || 0;

    try {
      const queryParams: Record<string, unknown> = { limit, offset };
      if (params.status) {
        queryParams.status = params.status;
      }

      const response = await this.httpClient.get<{
        proposals: Array<{
          id: number;
          title: string;
          description: string;
          creator: string;
          status?: string;
          created_at: string;
          voting_ends_at?: string;
          votes_for?: number;
          votes_against?: number;
          votes_abstain?: number;
        }>;
        total: number;
        limit: number;
        offset: number;
      }>('/governance/proposals', queryParams);

      const proposals = (response.proposals || []).map((p) => ({
        id: p.id,
        title: p.title,
        description: p.description,
        creator: p.creator,
        status: (p.status as ProposalStatus) || ProposalStatus.PENDING,
        createdAt: p.created_at,
        votingEndsAt: p.voting_ends_at,
        votesFor: p.votes_for || 0,
        votesAgainst: p.votes_against || 0,
        votesAbstain: p.votes_abstain || 0,
      }));

      return {
        data: proposals,
        total: response.total || 0,
        limit: response.limit || limit,
        offset: response.offset || offset,
      };
    } catch (error) {
      if (error instanceof GovernanceError) {
        throw error;
      }
      throw new GovernanceError(`Failed to list proposals: ${error}`);
    }
  }

  /**
   * Get proposal details
   *
   * @param proposalId - Proposal ID
   * @returns Proposal details
   *
   * @example
   * ```typescript
   * const proposal = await client.governance.getProposal(1);
   * console.log('Proposal:', proposal.title);
   * console.log('Votes for:', proposal.votesFor);
   * console.log('Votes against:', proposal.votesAgainst);
   * ```
   */
  async getProposal(proposalId: number): Promise<Proposal> {
    if (proposalId < 0) {
      throw new ValidationError('proposalId must be non-negative');
    }

    try {
      const response = await this.httpClient.get<{
        id: number;
        title: string;
        description: string;
        creator: string;
        status?: string;
        created_at: string;
        voting_starts_at?: string;
        voting_ends_at?: string;
        votes_for?: number;
        votes_against?: number;
        votes_abstain?: number;
      }>(`/governance/proposals/${proposalId}`);

      return {
        id: response.id,
        title: response.title,
        description: response.description,
        creator: response.creator,
        status: (response.status as ProposalStatus) || ProposalStatus.PENDING,
        createdAt: response.created_at,
        votingStartsAt: response.voting_starts_at,
        votingEndsAt: response.voting_ends_at,
        votesFor: response.votes_for || 0,
        votesAgainst: response.votes_against || 0,
        votesAbstain: response.votes_abstain || 0,
      };
    } catch (error) {
      if (error instanceof GovernanceError) {
        throw error;
      }
      throw new GovernanceError(`Failed to get proposal: ${error}`);
    }
  }

  /**
   * Create a governance proposal
   *
   * @param params - Proposal creation parameters
   * @returns Created proposal
   *
   * @example
   * ```typescript
   * const proposal = await client.governance.createProposal({
   *   title: 'Upgrade Protocol',
   *   description: 'Propose upgrading the protocol to v2.0',
   *   proposer: '0x1234...',
   *   duration: 604800 // 7 days
   * });
   * console.log('Proposal created:', proposal.id);
   * ```
   */
  async createProposal(params: CreateProposalParams): Promise<Proposal> {
    if (!params.title || !params.description || !params.proposer) {
      throw new ValidationError('title, description, and proposer are required');
    }

    try {
      const payload: Record<string, unknown> = {
        title: params.title,
        description: params.description,
        proposer: params.proposer,
      };

      if (params.duration) {
        payload.duration = params.duration;
      }
      if (params.metadata) {
        payload.metadata = params.metadata;
      }

      const response = await this.httpClient.post<{
        id: number;
        title: string;
        description: string;
        creator: string;
        status?: string;
        created_at: string;
        votes_for?: number;
        votes_against?: number;
      }>('/governance/proposals', payload);

      return {
        id: response.id,
        title: response.title,
        description: response.description,
        creator: response.creator,
        status: (response.status as ProposalStatus) || ProposalStatus.PENDING,
        createdAt: response.created_at,
        votesFor: response.votes_for || 0,
        votesAgainst: response.votes_against || 0,
      };
    } catch (error) {
      if (error instanceof GovernanceError) {
        throw error;
      }
      throw new GovernanceError(`Failed to create proposal: ${error}`);
    }
  }

  /**
   * Vote on a proposal
   *
   * @param proposalId - Proposal ID
   * @param voter - Voter address
   * @param choice - Vote choice (yes, no, abstain)
   * @returns Vote confirmation
   *
   * @example
   * ```typescript
   * const vote = await client.governance.vote(1, '0x1234...', VoteChoice.YES);
   * console.log('Vote submitted');
   * ```
   */
  async vote(
    proposalId: number,
    voter: string,
    choice: VoteChoice
  ): Promise<Record<string, unknown>> {
    if (proposalId < 0) {
      throw new ValidationError('proposalId must be non-negative');
    }

    if (!voter) {
      throw new ValidationError('voter is required');
    }

    if (![VoteChoice.YES, VoteChoice.NO, VoteChoice.ABSTAIN].includes(choice)) {
      throw new ValidationError("choice must be 'yes', 'no', or 'abstain'");
    }

    try {
      const payload = {
        voter,
        choice,
      };

      return await this.httpClient.post(`/governance/proposals/${proposalId}/vote`, payload);
    } catch (error) {
      if (error instanceof GovernanceError) {
        throw error;
      }
      throw new GovernanceError(`Failed to vote: ${error}`);
    }
  }

  /**
   * Get active proposals
   *
   * @returns List of active proposals
   *
   * @example
   * ```typescript
   * const activeProposals = await client.governance.getActiveProposals();
   * console.log('Active proposals:', activeProposals);
   * ```
   */
  async getActiveProposals(): Promise<Proposal[]> {
    try {
      const result = await this.listProposals({ status: 'active' });
      return result.data;
    } catch (error) {
      if (error instanceof GovernanceError) {
        throw error;
      }
      throw new GovernanceError(`Failed to get active proposals: ${error}`);
    }
  }

  /**
   * Get vote details for a proposal
   *
   * @param proposalId - Proposal ID
   * @returns Vote information
   *
   * @example
   * ```typescript
   * const votes = await client.governance.getProposalVotes(1);
   * console.log('Votes for:', votes.votesFor);
   * console.log('Votes against:', votes.votesAgainst);
   * console.log('Total votes:', votes.totalVotes);
   * ```
   */
  async getProposalVotes(proposalId: number): Promise<{
    proposalId: number;
    votesFor: number;
    votesAgainst: number;
    votesAbstain: number;
    totalVotes: number;
  }> {
    try {
      const proposal = await this.getProposal(proposalId);
      return {
        proposalId: proposal.id,
        votesFor: proposal.votesFor || 0,
        votesAgainst: proposal.votesAgainst || 0,
        votesAbstain: proposal.votesAbstain || 0,
        totalVotes:
          (proposal.votesFor || 0) +
          (proposal.votesAgainst || 0) +
          (proposal.votesAbstain || 0),
      };
    } catch (error) {
      if (error instanceof GovernanceError) {
        throw error;
      }
      throw new GovernanceError(`Failed to get proposal votes: ${error}`);
    }
  }
}

File: src/xai/sdk/python/xai_sdk/clients/governance_client.py
from __future__ import annotations

"""
Governance Client for XAI SDK

Handles governance and voting operations.
"""

from datetime import datetime
from typing import Any

from ..exceptions import GovernanceError, ValidationError
from ..http_client import HTTPClient
from ..models import Proposal, ProposalStatus

class GovernanceClient:
    """Client for governance operations."""

    def __init__(self, http_client: HTTPClient) -> None:
        """
        Initialize Governance Client.

        Args:
            http_client: HTTP client instance
        """
        self.http_client = http_client

    def list_proposals(
        self,
        status: str | None = None,
        limit: int = 20,
        offset: int = 0,
    ) -> dict[str, Any]:
        """
        List governance proposals.

        Args:
            status: Filter by proposal status
            limit: Number of proposals to retrieve
            offset: Offset for pagination

        Returns:
            List of proposals with metadata

        Raises:
            GovernanceError: If proposal list retrieval fails
        """
        if limit > 100:
            limit = 100

        try:
            params: dict[str, Any] = {"limit": limit, "offset": offset}
            if status:
                params["status"] = status

            response = self.http_client.get("/governance/proposals", params=params)

            proposals = [
                Proposal(
                    id=p["id"],
                    title=p["title"],
                    description=p["description"],
                    creator=p["creator"],
                    status=ProposalStatus(p.get("status", "pending")),
                    created_at=datetime.fromisoformat(p["created_at"]),
                    voting_ends_at=datetime.fromisoformat(p["voting_ends_at"])
                    if p.get("voting_ends_at")
                    else None,
                    votes_for=p.get("votes_for", 0),
                    votes_against=p.get("votes_against", 0),
                    votes_abstain=p.get("votes_abstain", 0),
                )
                for p in response.get("proposals", [])
            ]

            return {
                "proposals": proposals,
                "total": response.get("total", 0),
                "limit": response.get("limit", limit),
                "offset": response.get("offset", offset),
            }
        except GovernanceError:

            raise

        except (KeyError, ValueError, TypeError) as e:

            raise GovernanceError(f"Failed to list proposals: {str(e)}") from e

    def get_proposal(self, proposal_id: int) -> Proposal:
        """
        Get proposal details.

        Args:
            proposal_id: Proposal ID

        Returns:
            Proposal details

        Raises:
            GovernanceError: If proposal retrieval fails
        """
        if proposal_id < 0:
            raise ValidationError("proposal_id must be non-negative")

        try:
            response = self.http_client.get(f"/governance/proposals/{proposal_id}")

            return Proposal(
                id=response["id"],
                title=response["title"],
                description=response["description"],
                creator=response["creator"],
                status=ProposalStatus(response.get("status", "pending")),
                created_at=datetime.fromisoformat(response["created_at"]),
                voting_starts_at=datetime.fromisoformat(response["voting_starts_at"])
                if response.get("voting_starts_at")
                else None,
                voting_ends_at=datetime.fromisoformat(response["voting_ends_at"])
                if response.get("voting_ends_at")
                else None,
                votes_for=response.get("votes_for", 0),
                votes_against=response.get("votes_against", 0),
                votes_abstain=response.get("votes_abstain", 0),
            )
        except GovernanceError:

            raise

        except (KeyError, ValueError, TypeError) as e:

            raise GovernanceError(f"Failed to get proposal: {str(e)}") from e

    def create_proposal(
        self,
        title: str,
        description: str,
        proposer: str,
        duration: int | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> Proposal:
        """
        Create a governance proposal.

        Args:
            title: Proposal title
            description: Proposal description
            proposer: Address of proposer
            duration: Voting duration in seconds
            metadata: Optional metadata

        Returns:
            Created proposal

        Raises:
            GovernanceError: If proposal creation fails
        """
        if not title or not description or not proposer:
            raise ValidationError("title, description, and proposer are required")

        try:
            payload = {
                "title": title,
                "description": description,
                "proposer": proposer,
            }

            if duration:
                payload["duration"] = duration
            if metadata:
                payload["metadata"] = metadata

            response = self.http_client.post("/governance/proposals", data=payload)

            return Proposal(
                id=response["id"],
                title=response["title"],
                description=response["description"],
                creator=response["creator"],
                status=ProposalStatus(response.get("status", "pending")),
                created_at=datetime.fromisoformat(response["created_at"]),
                votes_for=response.get("votes_for", 0),
                votes_against=response.get("votes_against", 0),
            )
        except GovernanceError:
            raise
        except GovernanceError:

            raise

        except (KeyError, ValueError, TypeError) as e:

            raise GovernanceError(f"Failed to create proposal: {str(e)}") from e

    def vote(self, proposal_id: int, voter: str, choice: str) -> dict[str, Any]:
        """
        Vote on a proposal.

        Args:
            proposal_id: Proposal ID
            voter: Voter address
            choice: Vote choice (yes, no, abstain)

        Returns:
            Vote confirmation

        Raises:
            GovernanceError: If voting fails
        """
        if proposal_id < 0:
            raise ValidationError("proposal_id must be non-negative")

        if not voter:
            raise ValidationError("voter is required")

        if choice not in ["yes", "no", "abstain"]:
            raise ValidationError("choice must be 'yes', 'no', or 'abstain'")

        try:
            payload = {
                "voter": voter,
                "choice": choice,
            }

            return self.http_client.post(
                f"/governance/proposals/{proposal_id}/vote", data=payload
            )
        except GovernanceError:

            raise

        except (KeyError, ValueError, TypeError) as e:

            raise GovernanceError(f"Failed to vote: {str(e)}") from e

    def get_active_proposals(self) -> list[Proposal]:
        """
        Get active proposals.

        Returns:
            List of active proposals

        Raises:
            GovernanceError: If retrieval fails
        """
        try:
            result = self.list_proposals(status="active")
            return result["proposals"]
        except GovernanceError:

            raise

        except (KeyError, ValueError, TypeError) as e:

            raise GovernanceError(f"Failed to get active proposals: {str(e)}") from e

    def get_proposal_votes(self, proposal_id: int) -> dict[str, Any]:
        """
        Get vote details for a proposal.

        Args:
            proposal_id: Proposal ID

        Returns:
            Vote information

        Raises:
            GovernanceError: If retrieval fails
        """
        try:
            proposal = self.get_proposal(proposal_id)
            return {
                "proposal_id": proposal.id,
                "votes_for": proposal.votes_for,
                "votes_against": proposal.votes_against,
                "votes_abstain": proposal.votes_abstain,
                "total_votes": proposal.total_votes,
            }
        except GovernanceError:

            raise

        except (KeyError, ValueError, TypeError) as e:

            raise GovernanceError(f"Failed to get proposal votes: {str(e)}") from e
